compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -890,22 +875,56 @@ static void enforceLayoutToTransferWriteOp(
     return;
   }
 
-  // Build a transposed layout.
-  SmallVector<unsigned> permutation;
-  AffineMap permMap = write.getPermutationMap();
-  bool isSupportedPerm =
-      permMap.isPermutationOfMinorIdentityWithBroadcasting(permutation);
+  DistributionLayout *maskLattice = operandLattices[1];
+
   VectorLayoutInterface layout = writeOperand->getLayout();
-  SmallVector<int64_t> transposePerm(permutation.begin(), permutation.end());
-  if (isSupportedPerm) {
-    layout = layout.permute(transposePerm);
+  AffineMap maskMap =
+      inversePermutation(compressUnusedDims(write.getPermutationMap()));
+  VectorLayoutInterface maskLayout = layout.apply(maskMap);
+  ChangeResult changed = maskLattice->resolveWithPossibleConflict(
+      maskLayout, getOpOperand(write, 1));
+  update(maskLattice, changed);
+}
+
+static void enforceLayoutToTransferGatherOp(
+    TransferGatherOp gather, ArrayRef<DistributionLayout *> operandLattices,
+    ArrayRef<const DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
```

Comment:
nit: pass this as `llvm::function_ref` if you don't store the callback anywhere. The main difference is that `std::function` may dynamically allocate memory